src/core/logic.ml

```diff
@@ -927,6 +1518,756 @@ module Solver = struct
+
+
+       CHOICE:
+
+      1)   Q = Box(cPsi, A)      cD ; cG >> box(cPsihat.M) :Box(cPsi, A)
+
+             cD ; cPsi ==>  M: A     (this is solve!)
+            -----------------------------------------------------    [in Harpoon some msolve (args) ]
+               cD ; cG >> box(cPsihat.M) : Box(cPsi, A)                (assumes that cG is stable)
+
+
+      2) Q = Inductive(Atomic)
+
+         matchCompAtom (Signature) (a) = computation-level clauses c (for example Ev_z) that have a (for example Even) in its cHead
+         this will result in some subgoals (cgᵢ)
+
+         for all i.   cD ; cG ===>  eᵢ :  cgᵢ   (uniform right phase)
+         --------------------------------------------------------------------[in Harpoon some auto-solve]
+         cD ; cG >> c e₁ ...en  : Inductive(a, meta-spine)                       (assumes that cG is stable)
+
+
+      For example:
+
+      inductive Even: [ |- nat] -> ctype =
+      | Ev_z : Even [ |- z]
+      | Ev_ss: Pi N:[ |- nat]. Even [ |- N] -> Even [ |- suc (suc N)]
+
+
+    3) Q = Box(cPsi,A) or Inductive(a, meta-spine)  we solve it using some assumption from cG
+
+      cG might contain  f: tau₁ -> Q  , x:tau₁   ===> Q
+
+
+      focus left  (on some comp_goal cgᵢ in cG)
+
+        cD ; cGamma > f:cgᵢ ===> e : Q
+       --------------------------------- transition to focus left
+       cD ; cGamma  >> e : Q
+
+
+
+
+    cD ; cGamma > e e₁ : cg₂ ==>   Q     cD ; cGamma ===> e1 : cg₁
+    -----------------------------------------------------------------
+     cD ; cGamma > e : cg₁ -> cg₂ ==>   Q
+
+
+
+    cD ; cGamma > e C : [C/X]cg₂ ==>   Q
+    ----------------------------------------
+     cD ; cGamma > e : Pi X:U. cg₂ ==>   Q
+
+
+    ----------------------------------------
+    cD ; cGamma  > e : Q   ====> Q
+   *)
+
+  and solveSubGoals cD cG cPool (sg, k) ms mV sc fS currDepth maxDepth =
+    match sg with
+    | Proved ->
+       let e' = (Comp.Var (noLoc, k)) in
+       let e = fS e' in
+       let e = Whnf.cnormExp' (e, LF.MShift 0) in
+       sc e
+    | Solve (sg', cg') ->
+        (* printf "solve gamma SG \n";
+        let cg = normCompGoal (cg', ms) in
+        Printer.printState cD cG cg ms; *)
+       cgSolve' cD cG cPool (cg', ms)
+         (fun e ->
+           solveSubGoals cD cG cPool (sg', k) ms mV
+             (fun e' -> sc (Comp.Apply (noLoc, e', e))) fS currDepth maxDepth)
+         (succ currDepth) maxDepth
+
+
+  and solveCClauseSubGoals cD cG cPool cid sg ms mV sc fS currDepth maxDepth =
+   (* fprintf std_formatter "CID = %a \n"
+      Id.print (Index.compConstName cid); *)
+    match sg with
+    | Proved ->
+       let e' = (Comp.DataConst (noLoc, cid)) in
+       let e = fS e' in
+       let e = Whnf.cnormExp' (e, LF.MShift 0) in
+       sc e
+    | Solve (sg', cg') ->
+       (* printf "solve sig SG \n";
+       let cg = normCompGoal (cg', ms) in
+       Printer.printState cD cG cg ms; *)
```

Comment:
Debug print statements that could still be useful in the future may use the `Debug` module with `dprintf` declared at the top of `Logic` instead of being commented out.